### PR TITLE
fix: handle incoming requests on the api level

### PR DESCRIPTION
**Pull request by thomasheartman**
Update all endpoints of the frontend_api to accept top-level properties.

This is primarily just changing the type we expect and then, on the lower level functions, converting the incoming context into a regular context. 

Most of the functions are pretty straight-forward. `get_all_features` appears to be the only one that doesn't share an implementation with any other functions. 

The tests check:
- `/api/frontend/features/:feature` (`/api/proxy/features/:feature` uses the same impl under the hood, so I've not tested that explicitly. Happy to change that though.
- GET `/api/{frontend,proxy}(/all)?`
- POST `/api/{frontend,proxy)` (POST all features function takes an `IncomingContext` as a parameter, so I'm assuming the serialization here works; can of course add tests for that too)
- That providing a string as the value of `properties` yields a 400

Are there any more endpoints to update? Any other input?

### CODE REVIEW

File: server/src/frontend_api.rs [hunk 1]
```diff
@@ -1346,7 +1510,7 @@ mod tests {
             .set_json(json!({ "companyId": "bricks"}))
             .to_request();
         let result: FrontendResult = test::call_and_read_body_json(&app, req).await;
-        assert!(result.toggles.is_empty());
+        assert_eq!(result.toggles.len(), 1);
```
Comment (thomasheartman) [inline server/src/frontend_api.rs:1513]: Turns out we had a test that specifically checked that top-level properties don't work before. I've changed the assertion for now, but we could maybe also remove the last request and assertion?

OR we could treat this as verification for the post properties and remove the post test earlier in the file?

File: server/src/frontend_api.rs [hunk 2]
```diff
@@ -241,11 +241,12 @@ security(
 )
 )]
 #[get("")]
+#[instrument(skip(edge_token, req, engine_cache, token_cache))]
```
[the remark this answers is not in the document]
Comment (chriswk) [reply server/src/frontend_api.rs:244]: I think we might not want to include context either here, if we do, and we log, we'll have the entire struct exploded as context data in our log lines. Might expose things in logs we don't want to?
```suggestion
#[instrument(skip(edge_token, req, engine_cache, token_cache, context))]
```

File: server/src/frontend_api.rs (hunk 1 above)
Comment (chriswk) [reply server/src/frontend_api.rs:1520]: I'd say the earlier test is enough, feel free to remove this one.

**Review by chriswk — APPROVED** [review]
Great work, seems to cover all edge cases I could think of. :+1:

File: server/src/frontend_api.rs [hunk 3]
```diff
@@ -979,6 +1010,112 @@ mod tests {
...
+                    name: "disabled".into(),
+                    enabled: false,
+                    payload: None,
+                },
+                impression_data: false,
+            }],
+        };
+
+        assert_eq!(frontend_result, serde_json::to_vec(&expected).unwrap());
+    }
```
[the remark this answers is not in the document]
Comment (sighphyre) [reply server/src/frontend_api.rs]: Not a big fan of DRY tests, I tend to find one causes more damage to the clarity of what the test is doing. I don't really see this as a problem

File: server/src/frontend_api.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (thomasheartman) [reply server/src/frontend_api.rs:244]: Sure thing! Should I do the same for `get_enabled_frontend` too, so that they're the same?

File: server/src/frontend_api.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (thomasheartman) [reply server/src/frontend_api.rs]: All good! 💁🏼 I'll leave it. Thanks!

File: server/src/frontend_api.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (thomasheartman) [reply server/src/frontend_api.rs]: Done in 5276d209f07d3b7187c8b41cce3bbba956231720

**Comment by thomasheartman** [comment]
Ah, there is one edge case that I can think of, actually:
Deserializing empty string properties.

I noticed that the base Context type is defined with a custom deserializer: 

```
        deserialize_with = "remove_null_properties",
```

With the new extra_properties, these can be empty strings. Should they be stripped out in any way when deserializing?

(Also: added a test for the invalid properties value giving 400 as mentioned in the previous PR)
